db/collection.go

```diff
@@ -526,6 +529,283 @@ func validateUpdateSchemaFields(
+) error {
+	for _, newCol := range newColsByID {
+		oldCol, ok := oldColsByID[newCol.ID]
+		if !ok {
+			continue
+		}
+
+		// DeepEqual is temporary, as this validation is temporary
+		if !reflect.DeepEqual(oldCol.Indexes, newCol.Indexes) {
+			return NewErrCollectionIndexesCannotBeMutated(newCol.ID)
+		}
+	}
+
+	return nil
+}
+
+func validateFieldsNotModified(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		oldCol, ok := oldColsByID[newCol.ID]
+		if !ok {
+			continue
+		}
+
+		// DeepEqual is temporary, as this validation is temporary
+		if !reflect.DeepEqual(oldCol.Fields, newCol.Fields) {
+			return NewErrCollectionFieldsCannotBeMutated(newCol.ID)
+		}
+	}
+
+	return nil
+}
+
+func validateIDNotZero(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		if newCol.ID == 0 {
+			return ErrCollectionIDCannotBeZero
+		}
+	}
+
+	return nil
+}
+
+func validateIDUnique(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	colIds := map[uint32]struct{}{}
+	for _, newCol := range newColsByID {
+		if _, ok := colIds[newCol.ID]; ok {
+			return NewErrCollectionIDAlreadyExists(newCol.ID)
+		}
+		colIds[newCol.ID] = struct{}{}
+	}
+
+	return nil
+}
+
+func validateIDExists(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		if _, ok := oldColsByID[newCol.ID]; !ok {
+			return NewErrAddCollectionIDWithPatch(newCol.ID)
+		}
+	}
+
+	return nil
+}
+
+func validateRootIDNotMutated(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		oldCol, ok := oldColsByID[newCol.ID]
+		if !ok {
+			continue
+		}
+
+		if newCol.RootID != oldCol.RootID {
+			return NewErrCollectionRootIDCannotBeMutated(newCol.ID)
+		}
+	}
+
+	return nil
+}
+
+func validateSchemaVersionIDNotMutated(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+	for _, newCol := range newColsByID {
+		oldCol, ok := oldColsByID[newCol.ID]
+		if !ok {
+			continue
+		}
+
+		if newCol.SchemaVersionID != oldCol.SchemaVersionID {
+			return NewErrCollectionSchemaVersionIDCannotBeMutated(newCol.ID)
+		}
+	}
+
+	return nil
+}
+
+func validateCollectionNotRemoved(
+	oldColsByID map[uint32]client.CollectionDescription,
+	newColsByID map[uint32]client.CollectionDescription,
+) error {
+oldLoop:
+	for _, oldCol := range oldColsByID {
+		for _, newCol := range newColsByID {
+			// It is not enough to just match by the map index, in case the index does not pair
+			// up with the ID (this can happen if a user moves it)
```

Comment:
question: `moves it` moves what?

Reply:
The collection relative to it's index in the map, I'll replace `it`

- [ ] Replace `it`